language-plutus-core/docs/Builtins.md

```diff
@@ -0,0 +1,340 @@
+
+`Constant` now stores the tag of a type as well as a value of that that type.
+
+The original hardcoded builtins can now be recovered as
+
+```haskell
+data U a where
+    UInteger :: U Integer
+
+pattern Integer = SomeType UInteger
+pattern ConstantInteger = SomeValueIn UInteger
+
+type OriginalType = Type U
+type OriginalTerm = Term U
+```
+
+Now we can parameterize the AST by different universes of types. Of course, this requires support from all procedures (type checking, evaluation, etc) and every universe must implement a certain interface (see the [default universe](https://github.com/input-output-hk/plutus/blob/f9f81a441e29eeefc16546c4f3bc39e8614569eb/language-plutus-core/src/Language/PlutusCore/Universe/Default.hs) for an example).
+
+There are some technical difficulties on how to provide instances of various type classes for `SomeType` and `SomeValueIn`, how to make deriving work, etc (you can read about those in [source code of Plutus Core](https://github.com/input-output-hk/plutus/blob/f9f81a441e29eeefc16546c4f3bc39e8614569eb/language-plutus-core/src/Language/PlutusCore/Universe/Core.hs)), but otherwise this is pretty much the entire encoding.
+
+Implemented in [this PR](https://github.com/input-output-hk/plutus/pull/1849).
+
+## Extensible functions
+
+There are two approaches for making built-in functions extensible:
+
+1. instead of making `BuiltinFunction` an enumeration type, we can define it as
+
+```haskell
+newtype BuiltinFunction = BuiltinFunction Text
+```
+
+and provide a mapping from the name of a built-in function to its meaning, so that all relevant info is available during type checking and evaluation (we'll talk about meanings of builtins later)
+
+2. parameterize `Term` over `fun` representing a set of built-in functions. Unlike with universes we don't need to establish an isomorphism with some actual Haskell type right in the AST, so it's just the direct
+
+```haskell
+data Term fun
+    | AppBuiltinFunction fun [Term]
+    | <...>
+```
+
+and the original `Term` can be recovered by
+
+```haskell
+-- The original definition.
+data BuiltinFunction
+    = AddInteger
+    | SubtractInteger
+    | MultiplyInteger
+    | <...>
+
+type OriginalTerm = Term BuiltinFunction
+```
+
+Those two representations are very similar. Differences:
+
+- (2) requires adding another type variable, which can be annoying if you already have several of them (we do)
+- (2) is more expressive: it allows to specify at the type level what built-in functions a term may reference. Those can be built-in functions from a certain set (including the empty one) or you can instantiate `fun` with `Text` and get an equivalent of (1)
+- (1) requires adding another kind of errors: the name of a built-in function is not found in a mapping from built-in function names to their meanings during type checking or evaluation (not a big deal in practice)
+
```

Comment:
It'd be useful to have the names of all of the built-in functions/values available in the parser, so that it can immediately complain when it gets `[(builtin <some non-existent name>) arg1 ... argN]`.   Superficially that seems as if it'd be easier with (1) since you'd easily be able to extract such a list from the names -> meanings map.  The parser's not that important though, so that's not a big deal.  It currently doesn't check and if you give it a non-existent name then it fails during typechecking, or at runtime if the function isn't found.

Reply:
Good point, thanks, I'll mention that.